### ISSUE TITLE
🔥Internal error encountered FCM when sending notification to iOS

### ISSUE THREAD

**Issue by williamgoodhew** [invertase/react-native-firebase#1639]
## Issue

I am trying to send push notifications to an iOS app installed via TestFlight.

I can't get push notifications working with iOS react-native-firebase. The client is asking for permission correctly, and the pushToken is being stored correctly in the Firestore DB. I can send a notification to an Android device. I have tried to send notifications to iOS manually in the firebase console manually to a single device. I could not find any good logs but the status was: Failed - Invalid Team ID, Key ID and APNs Auth Key combination.

I tried sending a notification through cloud functions to get more detailed errors and I got the following error:

```
{ Error: Internal error encountered.
at FirebaseMessagingError.Error (native)
at FirebaseMessagingError.FirebaseError [as constructor] (/user_code/node_modules/firebase-admin/lib/utils/error.js:39:28)
at FirebaseMessagingError.PrefixedFirebaseError [as constructor] (/user_code/node_modules/firebase-admin/lib/utils/error.js:85:28)
at new FirebaseMessagingError (/user_code/node_modules/firebase-admin/lib/utils/error.js:241:16)
at Function.FirebaseMessagingError.fromServerError (/user_code/node_modules/firebase-admin/lib/utils/error.js:271:16)
at FirebaseMessagingRequestHandler.handleHttpError (/user_code/node_modules/firebase-admin/lib/messaging/messaging-api-request.js:125:50)
at /user_code/node_modules/firebase-admin/lib/messaging/messaging-api-request.js:113:23
at process._tickDomainCallback (internal/process/next_tick.js:135:7)
errorInfo: 
{ code: 'messaging/internal-error',
message: 'Internal error encountered.' },
codePrefix: 'messaging' }
```

here is the code for the cloud function which is working for Android but not on iOS:

```javascript
export const testNotification =
functions.https.onRequest((req, res) => {
  const message = {
		notification: {
			title: 'hello ios',
			body: "why won't you work",
		},
    token: 'czZaY1uryJs:APA91b***************'
	};


  return admin.messaging()
    .send(message)
    .then((response) => {
      console.log("Successfully sent message:", response);
    })
    .catch((error) => {
      console.log("Error sending message:", error);
    });
})
```

I have added some screenshots of the key set up that I have at the bottom of this post

---

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# Uncomment the next line to define a global platform for your project
platform :ios, '9.0'

target 'Athalens' do
  # Uncomment the next line if you're using Swift or would like to use dynamic frameworks
  # use_frameworks!
  pod 'Firebase'
  pod 'FirebaseCore'
  pod 'FirebaseAuth'
  pod 'FirebaseDatabase'
  pod 'FirebaseStorage'
  pod 'FirebaseFirestore'
  pod 'Firebase/Messaging', '~> 5.9.0'
  # Pods for Athalens
  
  target 'AthalensTests' do
    inherit! :search_paths
    # Pods for testing
  end

end

target 'Athalens-tvOS' do
  # Uncomment the next line if you're using Swift or would like to use dynamic frameworks
  # use_frameworks!

  # Pods for Athalens-tvOS


end
```

#### `AppDelegate.m`:

```objc
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "AppDelegate.h"
#import "RNFirebaseNotifications.h"
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
  [RNFirebaseNotifications configure];
  NSURL *jsCodeLocation;

  jsCodeLocation = [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];

  RCTRootView *rootView = [[RCTRootView alloc] initWithBundleURL:jsCodeLocation
                                                      moduleName:@"Athalens"
                                               initialProperties:nil
                                                   launchOptions:launchOptions];
  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  return YES;
}
  
  - (void)application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification {
    [[RNFirebaseNotifications instance] didReceiveLocalNotification:notification];
  }

@end
```

---

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

---

## Environment

- **Platform that you're experiencing the issue on**:
  - [X] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **If known, the version of the platform are you experiencing the issue on:**
  - `iOS 11.4.1`
- **Operating System:**
  - [ ] MacOS, version: `N/A`
  - [ ] Windows, version: `N/A`
  - [ ] Other, please specify: `N/A`
- **Build Tools:**
  - `Xcode 10`
- **`React Native` version:**
  - `0.56.0`
- **`React Native Firebase` library version:**
  - `4.3.8`
- **`Firebase` module(s) you're using that has the issue:**
  - [] **N/A**
  - [ ] Authentication
  - [ ] Analytics
  - [ ] Cloud **Firestore**
  - [x] Cloud **Messaging** (FCM)
  - [ ] Crashlytics
  - [ ] Dynamic **Links**
  - [ ] **Functions** Callable
  - [ ] Invites
  - [ ] Instance ID
  - [ ] Notifications
  - [ ] Performance Monitoring
  - [ ] Realtime **Database**
  - [ ] Remote **Config**
  - [ ] Storage
- **Are you using `TypeScript`?**
  - [ ] No
  - [X] Yes, version: `N/A`
- **Are you using Expo, e.g. `ExpoKit`?**
  - [x] No
  - [ ] Yes, I've _not_ ejected
  - [ ] Yes, but I **have** ejected to `ExpoKit`
  - [ ] Yes, but I **have** ejected to vanilla React Native
  - Expo version: `N/A`

<img width="618" alt="screenshot 2018-10-25 at 17 27 48" src="https://user-images.githubusercontent.com/32809621/47515816-bab80800-d87b-11e8-872b-3270aa482a5e.png">
<img width="695" alt="screenshot 2018-10-25 at 17 29 02" src="https://user-images.githubusercontent.com/32809621/47515817-bab80800-d87b-11e8-8df7-ff11e3b2b00f.png">
<img width="689" alt="screenshot 2018-10-25 at 17 30 11" src="https://user-images.githubusercontent.com/32809621/47515818-bab80800-d87b-11e8-9307-95c71f32ba6c.png">

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Donate via [Open Collective](https://opencollective.com/react-native-firebase/donate)
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter
- 👉 Star this repo on GitHub ⭐️
- 👉 Contribute; see our [contributing guide](./../../CONTRIBUTING.md)

**Comment by Aung-Myint-Thein** [comment]
Oooh.. is it a bug or feature request?

**Comment by williamgoodhew** [comment]
This is a bug.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Aung-Myint-Thein** [comment]
Oh does it mean we still cannot use react native firebase for iOS push notifications..? I am also getting the same error but this is first time implementation for me and I thought it was my mistake in somewhere.

**Comment by williamgoodhew** [comment]
@Salakar you have put an await-user-feedback tag on this - what feedback do you need from me?

**Comment by Aung-Myint-Thein** [comment]
I have managed to make this work by installing all the keys and certificates. Moreover, you can also try to send the push notification to the device from FCM API. It will provide more concrete and raw errors. For me, it says "InvalidAPNsCredentials" then I tried to set up the APN certificate again :)

**Comment by williamgoodhew** [comment]
hi @Salakar just wondering if you guys had any suggestions?

**Comment by Aung-Myint-Thein** [comment]
Ah, I downloaded a new set of provisional files for dev and distribution to run in xcode. Then, I downloaded a new set of APNSKey.

**Comment by gglee89** [comment]
@williamgoodhew I'm experiencing the exact same scenario.

I am able to get messages through FCM on **Android** but on **iOS** I get the `internal error` message.

```
{ Error: Internal error encountered.
    at FirebaseMessagingError.FirebaseError [as constructor] (/app/node_modules/firebase-admin/lib/utils/error.js:39:28)
    at FirebaseMessagingError.PrefixedFirebaseError [as constructor] (/app/node_modules/firebase-admin/lib/utils/error.js:85:28)
    at new FirebaseMessagingError (/app/node_modules/firebase-admin/lib/utils/error.js:241:16)
    at Function.FirebaseMessagingError.fromServerError (/app/node_modules/firebase-admin/lib/utils/error.js:271:16)
    at FirebaseMessagingRequestHandler.handleHttpError (/app/node_modules/firebase-admin/lib/messaging/messaging-api-request.js:125:50)
    at /app/node_modules/firebase-admin/lib/messaging/messaging-api-request.js:113:23
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
  errorInfo:
   { code: 'messaging/internal-error',
     message: 'Internal error encountered.' },
  codePrefix: 'messaging' }
```

**Comment by williamgoodhew** [comment]
I had an error in my bundle identifier which was only picked up using this package. My bundle identifier had a capital letter in accidentally, and that caused the error.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by gglee89** [comment]
@williamgoodhew Thank you for your comments.

I my case, I had my bundle identifier name all lowercased. So I could assume from the start that it was not what was generating the Firebase `internal-error`.

Fortunately I got it to work, but it took me few hours to figure out why.
It was a very dummy mistake, but **I forgot to upload my APN Authentication Key** to **Firebase Console**.

Just for the sake of having a visual representation:


And basically, it caused the unfriendly previously mentioned `internal-error`. Just to repeat:
```
{ code: 'messaging/internal-error',
     message: 'Internal error encountered.' },
  codePrefix: 'messaging' }
```

To have a better understanding of how iOS operates, I looked at the official video inside this post from Firebase:
https://firebase.googleblog.com/2017/01/debugging-firebase-cloud-messaging-on.html

Now that you gain a overall understanding of how iOS works, then I followed the instructions precisely as it is described in the official `react-native-firebase` documentation here: https://rnfirebase.io/docs/v5.x.x/messaging/ios

And Voilá. The sensitive point with iOS here is on how you organize and properly assign generated APN Certificates and Keys.

Hope it helps.
